### PR TITLE
Remove dropRight and takeRight from Iterator

### LINKED ISSUE #131: Remove `dropRight` from `Iterator`
(and same for methods that imply buffering on `Iterator`)

### PR DESCRIPTION
**Pull request by julienrf**
Fixes #131.

Note that I kept the `grouped` and `sliding` methods of `Iterator` even though they also use buffers. Since these methods do exist in 2.12 we have to discuss before we remove them…

### CODE REVIEW

File: src/main/scala/strawman/collection/View.scala [hunk 1]
```diff
@@ -113,14 +113,6 @@ object View extends IterableFactory[View] {
       if (underlying.knownSize >= 0) (underlying.knownSize - normN) max 0 else -1
   }
 
-  /** A view that drops trailing elements of the underlying collection. */
-  case class DropRight[A](underlying: Iterable[A], n: Int) extends View[A] {
-    def iterator() = underlying.iterator().dropRight(n)
-    protected val normN = n max 0
-    override def knownSize =
-      if (underlying.knownSize >= 0) (underlying.knownSize - normN) max 0 else -1
-  }
```
Comment (julienrf) [inline src/main/scala/strawman/collection/View.scala]: I removed the `DropRight` view which was based on the `dropRight` method of `Iterator`. (same for `TakeRight` below)

File: src/main/scala/strawman/collection/View.scala [hunk 2]
```diff
@@ -269,6 +253,8 @@ trait IndexedView[+A] extends View[A] with ArrayLike[A] { self =>
     }
   }
 
+  override def knownSize: Int = length
```
Comment (julienrf) [inline src/main/scala/strawman/collection/View.scala]: That’s unrelated to the PR, actually, but I thought we could override `knownSize` in the case of `IndexedView`. We should do the same for `IndexedSeq`.

File: src/main/scala/strawman/collection/View.scala [hunk 3]
```diff
@@ -286,11 +272,11 @@ object IndexedView {
...
 
-  class TakeRight[A](underlying: IndexedView[A], n: Int)
-  extends View.TakeRight(underlying, n) with IndexedView[A] {
-    override def iterator() = super.iterator() // needed to avoid "conflicting overrides" error
+  class TakeRight[A](underlying: IndexedView[A], n: Int) extends IndexedView[A] {
+    private val normN = n max 0
+    private val delta = underlying.length - normN
     def length = underlying.length min normN
-    def apply(i: Int) = underlying.apply(i)
+    def apply(i: Int) = underlying.apply(i + delta)
```
Comment (julienrf) [inline src/main/scala/strawman/collection/View.scala]: I kept the `IndexedView` based implementation of `takeRight` (and `dropRight`, below)

File: src/main/scala/strawman/collection/Iterable.scala [hunk 4]
```diff
@@ -475,7 +475,18 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
...
 
   /** A collection containing the last `n` elements of this collection. */
-  def takeRight(n: Int): C = fromSpecificIterable(View.TakeRight(coll, n))
+  def takeRight(n: Int): C = {
+    val b = newSpecificBuilder()
+    b.sizeHintBounded(n, coll)
+    val lead = coll.iterator() drop n
+    val it = coll.iterator()
+    while (lead.hasNext) {
+      lead.next()
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/Iterable.scala]: This is a cute algorithm, but I can't imagine it's faster than just going
```
var m = this.size
val it = coll.iterator()
while (m > n) it.next
while (it.hasNext) b += it.next()
```

File: src/main/scala/strawman/collection/Iterable.scala [hunk 5]
```diff
@@ -506,7 +517,17 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
   /** The rest of the collection without its `n` last elements. For
     *  linear, immutable collections this should avoid making a copy.
     */
-  def dropRight(n: Int): C = fromSpecificIterable(View.DropRight(coll, n))
+  def dropRight(n: Int): C = {
+    val b = newSpecificBuilder()
+    if (n >= 0) b.sizeHint(coll, -n)
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/Iterable.scala]: Why `-n`?

File: src/main/scala/strawman/collection/View.scala [hunk 6]
```diff
@@ -286,11 +272,11 @@ object IndexedView {
     def apply(i: Int) = underlying.apply(i)
   }
 
-  class TakeRight[A](underlying: IndexedView[A], n: Int)
-  extends View.TakeRight(underlying, n) with IndexedView[A] {
-    override def iterator() = super.iterator() // needed to avoid "conflicting overrides" error
+  class TakeRight[A](underlying: IndexedView[A], n: Int) extends IndexedView[A] {
+    private val normN = n max 0
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/View.scala]: You don't need to store this value.  `def length = underlying.length - delta` is enough.

File: src/main/scala/strawman/collection/View.scala [hunk 7]
```diff
@@ -300,16 +286,13 @@ object IndexedView {
     def apply(i: Int) = underlying.apply(i + normN)
   }
 
-  class DropRight[A](underlying: IndexedView[A], n: Int)
-  extends View.DropRight(underlying, n) with IndexedView[A] {
-    override def iterator() = super.iterator()
+  class DropRight[A](underlying: IndexedView[A], n: Int) extends IndexedView[A] {
+    private val normN = n max 0
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/View.scala]: You can just store the actual length here; you never need `n max 0` separately.

**Comment by Ichoran** [comment]
Looksreasonable.  I made some comments but didn't do a formal review.

File: src/main/scala/strawman/collection/Iterable.scala (hunk 4 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/Iterable.scala]: That would have the benefit of requiring only one traversal, indeed (excepted in the case of linear collections). I should give it a try.

File: src/main/scala/strawman/collection/Iterable.scala (hunk 5 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/Iterable.scala]: This version of `sizeHint` takes a collection and a delta to apply to the collection’s size. Here we want to say that the size of the result will be the size of `coll` minus the number of dropped elemnts. I could write explicitly `sizeHint(coll, delta = -n)`, for clarity.

File: src/main/scala/strawman/collection/Iterable.scala (hunk 4 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/Iterable.scala]: I tried it but didn’t notice a difference (in `List`):

takeright

Comment (Ichoran) [inline src/main/scala/strawman/collection/Iterable.scala]: Wait, what?  List shouldn't rebuild on `takeRight` anyway--if it does, that needs to be fixed.  Also, I didn't suggest the other method is faster, just more obviously correct.

Comment (julienrf) [inline src/main/scala/strawman/collection/Iterable.scala]: Indeed good point, that’s anyway something that we should fix…

**Pull request merged**